---
layout: default
title: Usage
---
<body>

<h1>Usage</h1>

<h2>Preliminary Information</h2>

<p>No matter how the user wishes to use FilTar, it is always invoked using the same format:</p>

<code>snakemake {name_of_target_file}</code>

<p>FilTar is run in this way, because it does not exist as a standard binary executable, or a script to be read by an interpreter, but rather as as a collection of snakefiles, scripts, text file and directory structures which relate to each other to achieve a common purpose i.e. the repository/directory itself is the tool.</p>

<p>As a result of this emphasis on directory structure, the user must ensure that they are located at the directory root when invoking snakemake.</p>

<p>Most parameters can be invoked by passing the <code>--config</code> option to snakemake and setting individual parameter values. Alternatively, the workflow can be configured by editing the <code>config/basic.yaml</code> file. See the 'configurations' page for more detail on this.

<h2>Standard Usage</h2>

<p>Standard and default usage refers to usage of the tool in which the user reannotates 3'UTRs, filters mRNA targets by expression level, and uses TargetScan7:</p>

<code>snakemake target_predictions.txt</code>

<p>Users can also optionally derive the genomic co-ordinates for identified predicted miRNA targets by running the following command: <code>snakemake {species}_target_predictions_with_gene_coords.txt</code></p>

<p>However, in order to run this step, users must ensure that the <code>reannotation</code> option is set to false when running this step. This step can also be used with target prediction data generated using the FilTarDB web application, however, they must ensure that the transcript name, miRNA name, target site 'start' and target site 'end' columns are named 'Gene ID','Mirbase ID','UTR start' and 'UTR end', respectively.</p>

<h2>Intermediate files</h2>

<p>Most intermediate files generated whilst FilTar is running are automatically deleted once the process is deleted. This prevents FilTar from returning identical output even when the user has changed their configuration options</p>

<p>Exceptions to this general rule are files which which will not change as a result of different user input such as raw genomic sequence and index files. Alignment and pseudoalignment intermediate files are also not deleted, as these only need to be updated if the user updates the number of raw sequencing files for a given <code>context</code>.</p>

One slight problem with this approach, is that it leads to some inefficiency as mostly intermediate files will have to be regenerated for every individual run of FilTar. For advanced users, with a good knowledge of both FilTar and the Snakemake job scheduling system, they can choose to edit FilTar source files to remove snakemake <code>temp</code> tags where they think it is appropriate to do so. However, caution is advised with this approach.

<h2>Modulating snakemake behaviour</h2>

<p>Many options can be passed to the snakemake command in order modulate behaviour, reference should be made to the official Snakemake <a href="https://snakemake.readthedocs.io/en/stable/">Snakemake documentation</a> as an exhaustive reference</p>

<p>Of particular importance, is to note that Snakemake has its own built-in job scheduling system to manage the execution of different rules. Many rules can be executed in parallel using the <code>--cores {num_cores}</code> option. Combined use of this option and execution within high-performance computing environments enable the execution of rules across many different cores.</p>

<h2>Warnings</h2>
<ol>
	<li><strong>Storage space:</strong> The script generating the context++ scores uses RNAplfold as a dependency which itself generates ~100KB of data per transcript per species per tissue. Therefore the option to use RNAplfold with FilTar is disabled unless this dependency is already present in the user's environment - this is to protect the user from an unintentional consumption of their own computational resources. If the user is aware of the risks, they can enable automated used of RNAplfold through conda by uncommenting the relevant conda directive for the rule generating context++ scores in <code>modules/target_prediction/targetscan/Snakefile</code> file. However, this will result in less information being used to score target predictions using targetscan.</li> 

	<li><strong>Cleaning output directories:</strong> Be cautious to delete files if running the same analysis (e.g. same species and tissue) using different 
3'UTR annotations, as targetscan7 will use RNAplfold generated from a previous analysis automatically. Again, manually deleting RNAplfold output will
protect against this issue.</li>
</ol>
</body>		
