#!/bin/bash

def get_bed_file(wildcards):
	if wildcards.species == 'hsa':
		return('results/Homo_sapiens.GRCh38.92.bed')
	elif wildcards.species == 'mmu':
		return('results/Mus_musculus.GRCm38.92.bed6')

def get_species(wildcards):
	if wildcards.species_prefix == 'Homo_sapiens.GRCh38':
		return('homo_sapiens')
	if wildcards.species_prefix == 'Mus_musculus.GRCm38':
		return('mus_musculus')

rule download_fetchChromSizes:
     output: "exe/fetchChromSizes"
     shell: "wget --directory-prefix=exe/ http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/fetchChromSizes"

rule download_gtfToGenePred:
     output: "exe/gtfToGenePred"
     shell: "wget --directory-prefix=exe/ http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/gtfToGenePred"

rule download_genePredToBed:
     output: "exe/genePredToBed"
     shell: "wget --directory-prefix=exe/ http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/genePredToBed"

rule get_gtf_file:
     output: "data/{species_prefix}.92.gtf"
     params: get_species
     shell: "wget --directory-prefix=data/ ftp://ftp.ensembl.org/pub/release-92/gtf/{params}/{wildcards.species_prefix}.92.gtf.gz && gunzip data/{wildcards.species_prefix}.92.gtf.gz && sed -i 's/^/chr/g' data/{wildcards.species_prefix}.92.gtf"

# derived from https://gist.github.com/gireeshkbogu/f478ad8495dca56545746cd391615b93

rule convert_gtf_to_genepred:
     input:
         script="exe/gtfToGenePred",
         gtf="data/Homo_sapiens.GRCh38.92.gtf"
     output:
         "results/Homo_sapiens.GRCh38.92.genePred"
     shell:
         "{input.script} {input.gtf} {output}"

rule convert_genepred_to_bed12:
     input:
         script="exe/genePredToBed",
         genepred="results/{species_prefix}.92.genePred"
     output:
         "results/{species_prefix}.92.bed"
     shell:
         "{input.script} {input.genepred} {output}"

rule get_chrom_sizes:
     input:
         "exe/fetchChromSizes"
     output:
         "results/{assembly}.chrom.sizes"
     shell:
         "{input} {wildcards.assembly} > results/{wildcards.assembly}.chrom.sizes"

rule get_bedgraph:
    input:
        sorted_bam="results/bam/{accession}.bam.sorted",
    output:
        "results/bam/run/{accession}.bedgraph"
    conda:
        "envs/bedtools.yaml"
    benchmark:
        "benchmarks/genomeCoverageBed/{accession}.txt"
    shell:
        "genomeCoverageBed -bg -ibam {input.sorted_bam} -split > {output}"

rule merge_bedgraphs_by_run:
        input:
             lambda wildcards: expand("results/bam/run/{accession}.bedgraph", accession=config['samples'][wildcards.sample])
        output:
             "results/bam/sample/{sample}_tmp.bedgraph"
        wildcard_constraints:
             sample="[E|D|S]RS.*"
        conda:
             "envs/bedtools.yaml"
        script: "merge_bedgraphs_by_sample.py"

rule avg_merged_bedgraph_by_run:
        input:
                "results/bam/sample/{sample}_tmp.bedgraph"
        output:
                "results/bam/sample/{sample}.bedgraph"
        wildcard_constraints:
             sample="[E|D|S]RS.*[0-9]"
        script:
                "get_average_bedgraph.R"

rule merge_bedgraphs_by_sample:
        input:
             lambda wildcards: expand("results/bam/sample/{sample}.bedgraph", sample=config['tissues'][wildcards.cell_line])
        output:
             "results/bam/tissue/{cell_line}.bedgraph.tmp"
        conda:
             "envs/bedtools.yaml"
        shell: "bedtools unionbedg -i {input} > {output}"

rule avg_merged_bedgraph_by_sample:
        input:
                "results/bam/tissue/{cell_line}.bedgraph.tmp"
        output:
                "results/bam/tissue/{cell_line}.bedgraph"
        script:
                "get_average_bedgraph.R"

rule reannotate_3utrs:
    input:
       script="exe/identifyDistal3UTR.pl",
       bed=get_bed_file,
       bedgraphs= 'results/bam/tissue/{cell_line}.bedgraph'
    benchmark: "benchmarks/APAtrap/{species}_{cell_line}.log"
    output:
       "results/bed/specific/{species}_{cell_line}.utr.bed"
    shell:
       "{input.script} -i {input.bedgraphs} -m {input.bed} -o {output}"

