---
layout: default
title: Configuration
---

<h1>Configuration:</h1>

<p>Default configurations can be set at the following configuration file: <code>config/basic.yaml</code>. Otherwise configurations can be set manually on a run-by-run basis when running FilTar from the command line by using the <code>--use-conda</code> option passed to snakemake</p>

<p>The primary options to be configured in the configuration files are as followed:</p>
<ol>
<li><code>tissues</code>: A mapping between tissue or cell line identifiers for a given species, their respective sample accession (c.f. <a href="https://www.ncbi.nlm.nih.gov/biosample/">BioSamples</a>). This mapping is mediated by a three letter species code (<i>e.g.</i> hsa) to differentiate between the same tissue type found in different species.</li> 
<li><code>samples</code>: The mapping of sample accessions to sequencing run accessions</li>
<li><code>single_end</code> and <code>paired_end</code>: A labelling of each run accession to determine if that library corresponds to single-end or paired-end sequencing</li>
<li><code>miRNAs</code>: A list of miRNAs (using canonical miRNA names found in miRBase with the three letter species prefix) to use for target prediction. If this configuration is left blank, target prediction is performed for all miRBase annotated miRNAs of that species</li>
<li><code>transcripts</code>: A list of mRNAs to use for target prediction. If this configuration is left blank, target prediction is performed for all annotated transcripts of a given species</li>
<li>The expression threshold on which context++ scores are filtered. To manage this, users should edit the 'params' directive of the results/no_reannotation or results/target_predication/targetscan Snakefiles for 'reannotation' and 'no reannotation' analyses respectively</li>
</ol>
<p>These details can be manually configured by editing the config/basic.yaml file to enter configurations in serialised format.</p>

<h2>User RNA-Seq data</h2>

<p>Users can provide their own data for use with FilTar</p>

<p>To do this, users must deposit their own compressed fastq data with <code>fastq.gz</code> or <code>fq.gz</code> file extension in either the <code>data/single_end</code> or <code>data/paired_end</code> directories. When using FilTar in this way, users must also set the <code>sequence_data_source</code> configuration value to <code>User</code></p>

<p>Files names must be in the form <code>{accession}.fastq.gz</code> for single-end data and <code>{accession}_1.fastq.gz</code> and <code>{accession}_2.fastq.gz</code> containing the first and second mate pair reads, respectively for paired-end data.</p> 

<p>Users can specify any value they wish for the accession wildcard provided that their corresponds to values specified in the main configuration file.</p>

<h2>Rule parameterisation</h2>

<p>For the most part, external tools are executed using default parameters which are documented in their respective snakefiles (found within subdirectories of the modules directory). Values of parameters can be altered by directly editing the 'param' key-value pairs contained within the rules of these snakefiles.</p>

