import os

def get_species_synonym(wildcards):
	if wildcards.species_prefix == 'Mus_musculus.GRCm38':
		return ('mus_musculus')
	elif wildcards.species_prefix == 'Homo_sapiens.GRCh38':
		return ('homo_sapiens')

def get_gtf_file(wildcards):
	if wildcards.species == 'mmu':
                return('data/Mus_musculus.GRCm38.92.gtf')
	elif wildcards.species == 'hsa':
		return('data/Homo_sapiens.GRCh38.92.gtf')

def get_assembly(wildcards):
	if wildcards.species == 'mmu':
		return("data/Mus_musculus.GRCm38.dna.primary_assembly.fa")
	if wildcards.species == 'hsa':
		return("data/Homo_sapiens.GRCh38.dna.primary_assembly.fa")

rule download_fasta:
        input:
        output: "data/{species_prefix}.dna.primary_assembly.fa.gz"
        params: get_species_synonym
        shell: "wget --directory-prefix=data/ ftp://ftp.ensembl.org/pub/release-92/fasta/{params}/dna/{wildcards.species_prefix}.dna.primary_assembly.fa.gz"

rule decompress_file:
        input: "data/{species_prefix}.dna.primary_assembly.fa.gz"
        output: "data/{species_prefix}.dna.primary_assembly.fa"
        threads: 8
        shell: "pigz -p {threads} -d {input}"

rule get_splice_sites_for_hisat2:
        input:
          gtf=get_gtf_file,
          py_script='exe/hisat2_extract_splice_sites.py'
        output: "results/{species}_hisat2_splice_sites.txt"
        shell: "python {input.py_script} {input.gtf} > {output}"

rule get_exons_for_hisat2:
        input:
          gtf=get_gtf_file,
          py_script='exe/hisat2_extract_exons.py'
        output: "results/{species}_hisat2_exons.txt"
        shell: "python {input.py_script} {input.gtf} > {output}"

rule create_index:
        input:
           assembly=get_assembly,
           splice_sites="results/{species}_hisat2_splice_sites.txt",
           exons="results/{species}_hisat2_exons.txt"
        conda:
           "envs/hisat2.yaml"
        output: "data/{species}.1.ht2"
        threads: 8
        shell: "hisat2-build -f -p {threads} --ss {input.splice_sites} --exon {input.exons} {input.assembly} data/{wildcards.species}"

def get_lib_type(wildcards):
	hisat2_dict = {'IU':'','ISF':'FR','ISR':'RF', 'U':'','SF':'F','SR':'R'}	
	#lib_type = config['lib_types'][wildcards.accession]
	if not os.path.exists('results/salmon/{}/lib_type.txt'.format(wildcards.accession)): return '' # stmps an error being generated when this file is missing
        fp=open('results/salmon/{}/lib_type.txt'.format(wildcards.accession))
        fp2 = fp.readlines()[0]
        lib_type = fp2.replace("\n","") 
	tmp = hisat2_dict[lib_type]
	
	if tmp == 'FR':
		return ("--rna-strandness FR")
	elif tmp == 'RF':
		return ("--rna-strandness RF")
	elif tmp == 'F':
		return ("--rna-strandness F")
	elif tmp == 'R':
		return ("--rna-strandness R")
	else:
		return ("")

def get_read_files (wildcards):
	if wildcards.accession in config['all_paired_end']:
		input_files = ['results/trimmed_fastq/{}_1_val_1.fq.gz'.format(wildcards.accession), 
			       'results/trimmed_fastq/{}_2_val_2.fq.gz'.format(wildcards.accession),
				] 
		return(input_files)
	else:
		input_file = ["results/trimmed_fastq/{}_trimmed.fq.gz".format(wildcards.accession)
                             ]
		return(input_file)

rule map_reads:
	input: 
		reads=get_read_files,
		lib_type_file='results/salmon/{accession}/lib_type.txt',
		index="data/mmu.1.ht2"
	conda:
		"envs/hisat2.yaml"
	params: get_lib_type
        output: "results/bam/{accession}.sam"
	script:
		"map_reads.py"

rule sam_to_bam:
	input: "results/bam/{accession}.sam",
        output: "results/bam/{accession}.bam"
        conda:
	   "envs/hisat2.yaml"
	shell: "samtools view -Sb  {input}  >  {output}"

rule samtools_sort:
    input:
        "results/{accession}.bam"
    output:
        "results/{accession}.bam.sorted"
    params:
        "-m 2G"
    threads: 1
    benchmark: "benchmarks/samtools_sort/{accession}.log"
    wrapper:
        "0.27.0/bio/samtools/sort"

#rule merge_bam:
#	input:
#		single_end = lambda wildcards: expand("/gpfs/afm/moxon/thomas2/APAtrap/results/bam/single_end/{accession}.bam.sorted", accession=config['sample_type']['human'][wildcards.tissue]['single-end']),
#		paired_end = lambda wildcards: expand("/gpfs/afm/moxon/thomas2/APAtrap/results/bam/paired_end/{accession}.bam.sorted", accession=config['sample_type']['human'][wildcards.tissue]['paired-end'])
#	output: "results/bam/{tissue}.bam"
#	shell: "samtools merge {output} {input.paired_end} {input.single_end}"

