def get_dna_file(wildcards):
        sci_species_name = config['sci_species_name'][wildcards.species]
        genome_build = config['genome_build'][wildcards.species]

        return('data/' + sci_species_name + '.' + genome_build + '.dna.chromosome.{}.fa'.format(wildcards.chrom))

rule fix_fasta_output:
	input: "results/msa/{species}_{tissue}_chr{chrom}_3UTR.tmp2.fa"
	output: "results/msa/{species}_{tissue}_chr{chrom}.fa"
	shell: "sed 's/(+)//g' {input} | sed 's/(-)//g' > {output}"

rule miRanda:
	input:
		utr="results/msa/{species}_{tissue}_chr{chrom}.fa",
		mirna="data/mirbase_mature_{species}.filtered.fa"
	output: "results/targets/miRanda/{species}/{tissue}_chr{chrom}.txt"
	conda: "envs/miRanda.yaml"
	shell: "miranda {input.mirna} {input.utr} > {output}"

rule convert_miRanda_to_tsv:
        input: "results/targets/miRanda/{species}/{tissue}_chr{chrom}.txt"
        output: "results/targets/miRanda/{species}/{tissue}_chr{chrom}.temp.tsv"
        shell: "scripts/convert_miRanda_to_tsv.sh {input} {output}"

rule aggregate_miRanda_output_no_reannotation:
        input: lambda wildcards: expand("results/targets/miRanda/{species}/{tissue}_chr{chrom}.temp.tsv", chrom=config['chromosomes'][wildcards.species], species=wildcards.species, tissue=wildcards.tissue)
        output: "results/targets/miRanda/{species}/{tissue}.temp.tsv"
        wildcard_constraints:
               tissue="((?!filt).)*"
        shell: "cat {input} > {output}"
